packages/react/src/utils/useScrollLock.ts

```diff
@@ -137,66 +137,89 @@ function preventScrollStandard(referenceElement: Element | null) {
   };
 }
 
+class ScrollLocker {
+  lockCount = 0;
+
+  restore = null as (() => void) | null;
+
+  unlockFrame = AnimationFrame.create();
+
+  acquire(referenceElement: Element | null) {
+    this.lockCount += 1;
+    if (this.lockCount === 1 && this.restore === null) {
+      this.lock(referenceElement);
+    }
+    return this.release;
+  }
+
+  release = () => {
+    this.lockCount -= 1;
+    this.unlockFrame.request(this.unlock);
+  };
+
+  private unlock = () => {
+    if (this.lockCount === 0 && this.restore) {
+      this.restore?.();
+      this.restore = null;
+    }
+  }
```

Comment:
Avoids unlock-lock sequences that create a lot of work. This is however a bit brittle, for example if React splits its work in multiple frames, the the unlock-lock sequence could happen and it would trigger reflows. We could instead use a timeout with a small delay to make it a bit more resilient. Wdyt?

![Screenshot From 2025-05-14 04-16-43](https://github.com/user-attachments/assets/3f21290d-ef79-4b8f-87c4-8115f5b231b5)